rust/etw_helpers/src/sessions.rs

```diff
@@ -0,0 +1,370 @@
+
+        let mut session_handle: CONTROLTRACE_HANDLE = Default::default();
+        let mut properties = EventTraceProperties::empty_for_query();
+        properties.set_session_name(&self.session_name);
+
+        let mut log_file_mode;
+
+        if let LogFileMode::File(file_mode) = self.log_file_mode {
+            match file_mode {
+                FileMode::Circular => {
+                    log_file_mode = EVENT_TRACE_FILE_MODE_CIRCULAR;
+                }
+                FileMode::Sequential => {
+                    log_file_mode = EVENT_TRACE_FILE_MODE_SEQUENTIAL;
+                }
+                FileMode::NewFile {
+                    maximum_file_size_in_kilobytes: kb,
+                } => {
+                    max_file_size = kb;
+                    log_file_mode = EVENT_TRACE_FILE_MODE_NEWFILE;
+                }
+            }
+
+            properties.set_file_name(&self.file_name);
+
+            if self.realtime {
+                log_file_mode |= EVENT_TRACE_REAL_TIME_MODE;
+            }
+        } else {
+            log_file_mode = EVENT_TRACE_BUFFERING_MODE;
+            properties.props.LogFileNameOffset = 0;
+        }
+
+        properties.props = EVENT_TRACE_PROPERTIES {
+            Wnode: WNODE_HEADER {
+                ClientContext: 1,
+
+                ..properties.props.Wnode
+            },
+            BufferSize: self.buffer_size as u32,
+            MinimumBuffers: self.min_buffers as u32,
+            MaximumBuffers: self.max_buffers as u32,
+            LogFileMode: log_file_mode,
+            NumberOfBuffers: self.num_buffers as u32,
+            FlushTimer: 1,
+            MaximumFileSize: max_file_size,
+
+            ..properties.props
+        };
+
+        unsafe {
+            if recreate_existing_session {
+                let existing_session =
+                    EtwSession::get_etw_session(PCSTR(properties.session_name.as_ptr()));
+                if existing_session.is_ok() {
+                    // Stop existing session
+                    drop(existing_session);
+                } else if existing_session.is_err()
+                    && existing_session.as_ref().unwrap_err_unchecked().code()
+                        == crate::error::ETW_SESSION_NOT_FOUND
+                {
+                    // No existing session
+                } else {
+                    // There's be an error, but ignore it and use the error StartTrace returns instead
+                }
+            }
+
+            let ptr = &mut properties.props as *mut EVENT_TRACE_PROPERTIES;
+            let err = StartTraceA(
+                &mut session_handle,
+                PCSTR(properties.session_name.as_ptr()),
+                ptr,
+            );
+
+            if err.is_err() {
+                Err(err.into())
+            } else {
+                Ok(ControlTraceHandle(session_handle))
+            }
+        }
+    }
+}
+
+impl ControlTraceHandle {
+    pub fn manual_stop(self) -> ManuallyDrop<Self> {
```

Comment:
Manual dropping seems unnecessary here, and is confusing. In addition, not all users of this API are going to want their sessions to stop when the handle goes out of scope -- it might be ok to make it the default, but it isn't always going to be the case.

What if instead we had a "stop_on_drop" field that defaulted to true, and could be set to false (either explicitly or by the call to manual_stop)?